[c10d] Enforce order of waited ranks in monitored barrier.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55265 [c10d] Log API usage of monitored barrier
* #55197 [c10d] monitored_barrier: ensure all ranks pass or none do
* #55010 [c10d] Monitored barrier: option to collect all failed ranks
* **#55009 [c10d] Enforce order of waited ranks in monitored barrier.**

Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

rohan-varma committed:
[c10d] Enforce order of waited ranks in monitored barrier.

Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

[ghstack-poisoned]

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5233,3 +5235,20 @@ def test_monitored_barrier_gloo_rank_0_timeout(self):
+        def test_monitored_barrier_failure_order(self):
+            # Ensure that the first (in sorted order) rank is reported when
+            # multiple ranks fail to pass the monitored_barrier.
+            # TODO: Provide ability to wait and report all failed ranks:
```
Nit: You can use `TODO(#54879)` for simplicity.

rohan-varma committed:
Update on "[c10d] Enforce order of waited ranks in monitored barrier."


Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Enforce order of waited ranks in monitored barrier."


Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Enforce order of waited ranks in monitored barrier."


Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Enforce order of waited ranks in monitored barrier."


Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Enforce order of waited ranks in monitored barrier."


Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Enforce order of waited ranks in monitored barrier."


Changes monitoredBarrier so that we await acknowledgemenet from ranks
in a consistent order (from least to greatest). This will reduce confusion
around the order the ranks are awaited. We are still planning to add support
for awaiting all ranks in follow up changes.

Differential Revision: [D27405417](https://our.internmc.facebook.com/intern/diff/D27405417/)

[ghstack-poisoned]
